ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/KsqlResource.java

```diff
@@ -869,47 +913,48 @@ private void validateShowColumns(final PreparedStatement<ShowColumns> statement)
       if (showColumns.isTopic()) {
         final KsqlTopic ksqlTopic = executionSandbox.getMetaStore().getTopic(name);
         if (ksqlTopic == null) {
-          throw new KsqlException(String.format(
-              "Could not find Topic '%s' in the Metastore",
-              name
-          ));
+          throw new KsqlStatementException(
+              "Could not find Topic '" + name + "' in the Metastore",
+              statement.getStatementText());
         }
       } else {
         final StructuredDataSource dataSource = executionSandbox.getMetaStore().getSource(name);
         if (dataSource == null) {
-          throw new KsqlException(String.format(
-              "Could not find STREAM/TABLE '%s' in the Metastore",
-              name
-          ));
+          throw new KsqlStatementException(
+              "Could not find STREAM/TABLE '" + name + "' in the Metastore",
+              statement.getStatementText());
         }
       }
     }
 
     private void validateExplain(final PreparedStatement<Explain> statement) {
-      final String queryId = statement.getStatement().getQueryId();
-
-      if (queryId == null) {
-        explainStatement(
-            statement.getStatement().getStatement(),
-            statement.getStatementText().substring("EXPLAIN ".length()),
-            executionSandbox,
-            ksqlConfig,
-            scopedPropertyOverrides);
-      } else {
-        explainQuery(queryId, executionSandbox);
-      }
+      explain(statement, scopedPropertyOverrides, ksqlConfig, executionSandbox);
     }
 
     private void validateDescribeFunction(final PreparedStatement<DescribeFunction> statement) {
-      final String functionName = statement.getStatement().getFunctionName();
+      try {
+        final String functionName = statement.getStatement().getFunctionName();
 
-      final FunctionRegistry functionRegistry = executionSandbox.getMetaStore();
-      if (!functionRegistry.isAggregate(functionName)) {
-        // No a known UDAF, see if know UDF. (The below throws on unknown method).
-        functionRegistry.getUdfFactory(functionName);
+        final FunctionRegistry functionRegistry = executionSandbox.getMetaStore();
+        if (!functionRegistry.isAggregate(functionName)) {
+          // Not a known UDAF, see if know UDF. (The below throws on unknown method).
+          functionRegistry.getUdfFactory(functionName);
+        }
+      } catch (final KsqlException e) {
+        throw new KsqlStatementException(e.getMessage(), statement.getStatementText(), e);
       }
     }
 
+    private void validateTerminateQuery(final PreparedStatement<TerminateQuery> statement) {
```

Comment:
with the new mode of working we can now correctly validate `TERMINATE` statements, where previously this was not possible.